### ISSUE TITLE
Adding "Curly Vector" Options With Streamplot or Quiver Plots

### ISSUE THREAD

**Issue by michaelavs** [matplotlib/matplotlib#20038]
### Problem

I am currently working to translate an older coding language into Python and have had a user ask about a "Curly Vector" function. The function is very similar to what is already in matplotlib in Streamplot and Quiver plots, but will take the curved nature of the arrow shaft from Streamline and the "disconnected" vector aspect from Quiver. This function would be very useful for visualizing flow patterns like those of ocean currents or wind. 

### Proposed Solution

I propose adding a feature to streamplot that will allow for vectors to be separate, but still curved in nature. I have seen a fix for this that essentially modifies the matplotlib code to combine the two features, but it does not work for the dataset I am using. The code can be found [here](https://github.com/kieranmrhunt/curved-quivers) and a link to the original stackoverflow question is [here](https://stackoverflow.com/questions/51843313/flow-visualisation-in-python-using-curved-path-following-vectors). 

### Additional context and prior art

A visual example of a Curly Vector with its current matplotlib/python recreation:

Original
![image](https://user-images.githubusercontent.com/43652875/115610735-7ff58600-a2a6-11eb-909d-bbdb74250b91.png)

Matplotlib/Python
![image](https://user-images.githubusercontent.com/43652875/115610769-8be14800-a2a6-11eb-8060-56fcb7035b4b.png)

**Comment by jklymak** [comment]
I think you would need a different API than either quiver or streamplot though you'd perhaps want to use the stream plot integration code.  This all seems pretty standalone, but if you had a working implementation, I would think it could be considered.

**Comment by jklymak** [comment]
Ooops, I should have kept reading on stackoverflow.  [This answer](https://stackoverflow.com/a/65607512/3394386) is basically what you should do.  Maybe it deserves an example in the gallery, but it looks like the functionality you want is there?    (though I guess I'm not 100% sure if that answer scales the length of the arrows properly)

**Comment by michaelavs** [comment]
> Ooops, I should have kept reading on stackoverflow. [This answer](https://stackoverflow.com/a/65607512/3394386) is basically what you should do. Maybe it deserves an example in the gallery, but it looks like the functionality you want is there? (though I guess I'm not 100% sure if that answer scales the length of the arrows properly)

I've tried the options on that stackoverflow, however they do not work with the data set used to create the plots I have above (which originally use plt.quiver) and on a separate plot that uses plt.streamplot(). I have also found that the one you specifically linked, does not actually run as it is currently written; there is a RuntimeWarning and no plot is displayed. I do agree that the example you have linked is the same functionality that I am asking about, however, I think this issue may end up larger than just adding an example to the metplotlib website. From the most [popular answer](https://stackoverflow.com/a/52000829) in the discussion, it would appear that the correct method involves actually changing the source code from matplotlib and creating a new function to plot with.

**Comment by jklymak** [comment]
If I were interested in doing this, I would do as a standalone package initially, and if it really seems useful, pitch it as an addition.  I personally feel this is a bit specialized for core, but maybe if its not too much extra code and the API is simple it could work....

**Comment by veenstrajelmer** [comment]
@jklymak: I have been following several curved quiver discussions in the past. Would it be possible to add the [curved-quivers](https://github.com/kieranmrhunt/curved-quivers) code from @kieranmrhunt to matplotlib? It is also linked [in his answer](https://stackoverflow.com/a/52000829/5578371) on stackoverflow discussion you followed.

**Comment by jklymak** [comment]
I can't quite tell how well architected that package is.   Would it not be better listed at mpl-third-party (https://matplotlib.org/mpl-third-party/) and the community continue to work on it?  Our quiver and streamline code are already somewhat brittle, so adding another plot type seems a heavy lift.

**Comment by veenstrajelmer** [comment]
@jklymak I agree, further testing it gave many unexpected errors. In that case I think it would be great to have a bit more control over the plt.streamlines(), like providing length and put the arrow on the front instead of half-way. At least something to make it a bit more towards quiver. The answer you are pointing to (https://github.com/matplotlib/matplotlib/issues/20038#issuecomment-824324310) does exaclty but is not generic and it includes a for loop which becomes slow for large datasets.

Here is some example code with xarray testdata:
```python
import numpy as np
import matplotlib.pyplot as plt
import xarray as xr

ds = xr.tutorial.load_dataset("eraint_uvz").isel(month=0,level=0)
ds = ds.sortby('latitude') #.sortby() to avoid "ValueError: 'y' must be strictly increasing"
x = ds.longitude.to_numpy()
y = ds.latitude.to_numpy()
U = ds.u.to_numpy()
V = ds.v.to_numpy()
speed = np.sqrt(U**2 + V**2)

fig, ax = plt.subplots()
quiv_curved = ax.streamplot(x,y,U,V,color=speed, arrowstyle='-|>', integration_direction='backward',broken_streamlines=True, density=2)
```

Is it possible to put the arrow at the front and to steer the length of the lines? My tries with `minlength` and `maxlength` resulted in an almost empty plot. Maybe you are right and it should be a third party community effort, but it seems so useful to me to be able to make nice curved quiver plots.

**Comment by veenstrajelmer** [comment]
Ok, I dived into it. I have aligned the function by @kieranmrhunt with matplotlib.streamplot and simplified some things to make merging easier. The code is available here: https://github.com/Deltares/dfm_tools/blob/main/dfm_tools/modplot.py
This code compares quite well to [matplotlib.streamplot](https://raw.githubusercontent.com/matplotlib/matplotlib/main/lib/matplotlib/streamplot.py).

Some points of attention, when looking at the differences:
- to me it makes more sense to use `grains` (n-points in x/y direction) than `density` (1 gives 30x30 raster). There is also overlap with 'scale' and probably `10` in `mask = StreamMask(10)` should be replaced with `grains` or something similar.
- `minlength` and `maxlength` are not given as arguments, but instead `magnitude` and `resolution` are computed, the latter with `scale` and `grains`. Some design choice has to be make in order to merge it.
- `start_points` are instantly generated with a mesh instead of rotation wise. The effect on the normal streamplot is visible, but it still quite good and the method is simpler. The other way round, I could not get it to work.
- arrows are put in middle (`n = np.searchsorted(s, s[-1] / 2.)`) vs front (`n = np.searchsorted(s, s[-1])`), add `arrowloc` keyword to make both possible. And maybe change default to front of arrow?

@jklymak could you maybe have a look at it and let me know how you see above points? I think it is relatively easy to make a PR now, but some input is welcome.

**Comment by jklymak** [comment]
I think my vote would still be that this would be a separate package.  Unless there is something this needs that means it needs to reach into Matplotlib's guts, supporting this seems out of scope for us.

**Comment by veenstrajelmer** [comment]
Ok, that's too bad. I would have expected it be mergeable with streamplot since the code is almost the same code. I will just keep this adapted copy of streamplot in my own code for now, since it is a bit too much of an effort to create and maintain a separate package for one script.

**Comment by cvanelteren** [comment]
We are interested in adding it to [UltraPlot](https://github.com/Ultraplot/UltraPlot) @veenstrajelmer

**Comment by veenstrajelmer** [comment]
I just checked it out in Ultraplot and it works quite nice: https://github.com/Ultraplot/UltraPlot/discussions/370

Code snipped:
```python
# Source - https://stackoverflow.com/a/75894627
# Posted by Jelmer Veenstra, modified by community. See post 'Timeline' for change history
# Retrieved 2026-04-07, License - CC BY-SA 4.0

import numpy as np
import matplotlib.pyplot as plt
plt.close('all')
import ultraplot

x = np.linspace(-4,4,120)
y = np.linspace(-3,3,100)
X,Y = np.meshgrid(x,y)
U = -1 - X**2 + Y
V = 1 + X - Y**2
speed = np.sqrt(U*U + V*V)

fig,(ax1,ax2) = plt.subplots(1,2,figsize=(10,5),sharex=True,sharey=True)

# streamplot
strm = ax1.streamplot(X, Y, U, V, color=speed, cmap='winter', arrowstyle='fancy', linewidth=speed/5, integration_direction='both')
fig.colorbar(strm.lines)
ax1.set_title("matplotlib streamplot")

# curved_quiver
strm = ultraplot.axes.PlotAxes.curved_quiver(
    self=ax2, x=X, y=Y, u=U, v=V, color=speed, cmap='winter',
    linewidth=speed/5,
    arrow_at_end=True,
    density=5,
    scale=5,
    arrowsize=1,
    grains=15,
    )
fig.colorbar(strm.lines)
ax2.set_title("UltraPlot curved_quiver")
```
Gives:

<img width="956" height="486" alt="Image" src="https://github.com/user-attachments/assets/a75fa9bb-e3ac-4ec6-8ddd-904f8e9e6c18" />

**Comment by cvanelteren** [comment]
Small aside: it is easier to plot natively both on our axes ;-)


```python

import numpy as np
import ultraplot as uplt

x = np.linspace(-4,4,120)
y = np.linspace(-3,3,100)
X,Y = np.meshgrid(x,y)
U = -1 - X**2 + Y
V = 1 + X - Y**2
speed = np.sqrt(U*U + V*V)

fig,(ax1,ax2) = uplt.subplots(ncols = 2,figsize=(10,5))

# streamplot
strm = ax1.streamplot(X, Y, U, V, color=speed, cmap='winter', arrowstyle='fancy', linewidth=speed/5, integration_direction='both')
fig.colorbar(strm.lines)
ax1.set_title("matplotlib streamplot")

# curved_quiver
strm = ax2.curved_quiver(
    x=X, y=Y, u=U, v=V, color=speed, cmap='winter',
    linewidth=speed/5,
    arrow_at_end=True,
    density=5,
    scale=5,
    arrowsize=1,
    grains=15,
    )
ax2.set_title("UltraPlot curved_quiver")
```
